src/security/core/src/dds_security_plugins.c

```diff
@@ -211,7 +211,8 @@ dds_return_t dds_security_load_security_library (const dds_security_plugin_confi
   if (lib_ret != DDS_RETCODE_OK)
   {
     char buffer[256];
-    ddsrt_dlerror (buffer, sizeof (buffer));
+    if (ddsrt_dlerror (buffer, sizeof (buffer)) != DDS_RETCODE_OK)
+      (void) snprintf (buffer, sizeof (buffer), "unknown error");
```

Comment:
I appreciate the attempt, but I looking at the code for `ddsrt_dlopen` and `ddsrt_dlerror`, I think the real fix is ditching the `dlerror` wrapper entirely and just return the proper return code from `ddsrt_dlopen`. A quick-fix for now could be printing "unknown error" to the buffer in `ddsrt_dlerror` or something, returning `DDS_RETCODE_NOT_FOUND` from `ddsrt_dlerror` makes no sense imho

Reply:
@dennis-adlink, you asked me what I meant with this comment offline, so I'll elaborate. To be honest `ddsrt_dlopen` and `ddsrt_dlclose` are both so-so, I haven't even looked at the other functions yet. For instance, `ddsrt_dlerror` uses an `snprintf`-like signature and actually uses `snprintf` to print to the user provided buffer. There's at least a couple things wrong with that:
1. What if I provide `0` for `buflen`, I'm pretty sure that's undefined behaviour for some platforms
2. If I provide a `buflen` of `1`, that just gets me an empty string, so we must check for the number of bytes required
3. We have a perfectly fine `ddsrt_strlcpy`, no need for "formatting" the string
4. `DDS_RETCODE_NOT_FOUND` just does not make any sense as an error code. The function just shouldn't fail and should print `"unkown error"` itself and return the number of bytes required

So your fix, fixes it for this instance, but next time someone else will make the same mistake. Even if `ddsrt_dlclose` is "successful", the contents (as far as the function signature goes), may contain a completely unusable message.

imho, it would be better to return `DDS_RETCODE_ERROR` from `ddsrt_dlopen` and simply remove `ddsrt_dlclose`, much easier. I should mention that `ddsrt_dlopen` is badly implemented too. Take for instance:

```
  if ((translate) && (strrchr(name, '/') == NULL )) {
    /* Add lib and suffix to the name and try to open. */
#if __APPLE__
    static const char suffix[] = ".dylib";
#else
    static const char suffix[] = ".so";
#endif
    char* libName;
    ddsrt_asprintf( &libName, "lib%s%s", name, suffix);
    *handle = dlopen(libName, RTLD_GLOBAL | RTLD_NOW);
    ddsrt_free(libName);
  }
```
1. Translation is probably done automatically by the linker to begin with
2. If it isn't, wouldn't it make more sense to check if the name ends with `.dylib` or `.so` instead of checking if the path contains a `/`. What if the library is located in the current directory?
3. The return code of `ddsrt_asprintf` isn't checked, so `libName` might just be uninitialized.
3.1. Possibly resulting in passing an uninitialized variable to `dlopen`; and
3.2. Passing an uninitialized pointer to `ddsrt_free`

I don't require you to fix these issues (you didn't introduce them), so I'll take your fix. But if you can spare the time and feel like really fixing it, that'd be even better.

Reply:
@k0ekk0ek thanks for your feedback! 

1-3 regarding `ddsrt_dlerror`: I've refactored the code based on your suggestions (db09700). 

4: I still think an error return code makes sense here: if someone calls the `ddsrt_dlerror` function in case no error related to dynamic library loading occurred, this should return something else than the message 'unknown error'. In `dds_security_plugins.c:215` we know that something went wrong and there should be an error, so that's why I use the 'unknown error' at that point. I've replace the error code with 'PRECONDITION_NOT_MET', which better describes what happened imho. Removing `ddsrt_dlerror` (you wrote `ddsrt_dlclose`, but I assume you meant `ddsrt_dlerror`) completely would make the log messages for library loading less clear: currently it e.g. reports 'library file not found' or 'library file corrupted'. Without the call to `dlerror` we'd not have this information. 

Regarding `ddsrt_dlopen` 1-2: the check is 
"translate == true and not a [absolute/relative] path in the `name` parameter" 
(it checks for _not_ a slash in `name`). E.g. the value of `name` is just something like `dds_security_ac`, which is not correctly handled by dlopen, which requires the full file-name as far as I understand from the manual. So I think the code is correct and useful; i.e. you don't have to add the file extension and `lib` prefix and can have a portable value in configuration when the library is just referenced by its name (which will be used in most cases I think, having the (library) path set correctly). 
3: you're right, I've refactored the code a bit

Reply:
Some additional refactoring of the dynamic library loading error handling is in commit d6bd655: on Windows it's now using a thread local variable to store the last error, so that the behavior of `ddsrt_dlerror` on Windows is now is equal to the posix and other variants. And this function now returns the number of characters written in the error message (0 in case of no error), or `DDS_RETCODE_NOT_ENOUGH_SPACE` in case the message does not fit in.